include/cuda/std/detail/libcxx/include/atomic

```diff
@@ -1243,269 +1243,282 @@ _LIBCUDACXX_INLINE_VISIBILITY void __cxx_atomic_wait(__cxx_atomic_impl<_Tp, _Sco
         __cxx_atomic_try_wait_slow(__a, __val, __order);
 }
 
-// general atomic<T>/atomic_ref<T>
-template <class _Tp, int _Sco = 0, bool = (is_integral<_Tp>::value || is_floating_point<_Tp>::value) && !is_same<_Tp, bool>::value>
-struct __atomic_base {
-    mutable __cxx_atomic_impl<_Tp, _Sco> __a_;
+template <class _Tp, typename _Storage>
+struct __atomic_base_storage {
+    mutable _Storage __a_;
 
-    __atomic_base() = default;
-    __atomic_base(const __atomic_base&) = delete;
-    __atomic_base(__atomic_base&&) = delete;
+    __atomic_base_storage() = default;
+    __atomic_base_storage(const __atomic_base_storage&) = default;
+    __atomic_base_storage(__atomic_base_storage&&) = default;
 
-    __atomic_base& operator=(const __atomic_base&) = delete;
-    __atomic_base& operator=(__atomic_base&&) = delete;
+    __atomic_base_storage& operator=(const __atomic_base_storage&) = default;
+    __atomic_base_storage& operator=(__atomic_base_storage&&) = default;
 
     _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR
-    __atomic_base(const _Tp& __a) _NOEXCEPT : __a_(__a) {}
+    __atomic_base_storage(_Storage&& __a) _NOEXCEPT : __a_(forward<_Storage>(__a)) {}
+};
+
+template <class _Tp, bool _Cq, typename _Storage>
```

Comment:
Could we get a more descriptive name instead of `_Cq`?

Reply:
I'll push an update with `_Cq`->`_ConstQualified`